stdlib/src/builtin/string.mojo

```diff
@@ -1979,6 +1980,101 @@ struct String(
             )
         return String(buf^)
 
+    @always_inline
+    fn format[*Ts: Stringable](self, *args: *Ts) raises -> String:
+        """Format a template with *args.
+
+        Example of manual indexing:
+
+        ```mojo
+        print(
+            String("{0} {1} {0}").format(
+                "Mojo", 1.125
+            )
+        ) #Mojo 1.125 Mojo
+        ```
+
+        Example of automatic indexing:
+
+        ```mojo
+        var x = String("{} {}").format(
+            True, "hello world"
+        )
+        print(x) #True hello world
+        ```
+
+        ⚠️ Raising functions does not work in the parameter domain.
+
+        Args:
+            args: Values passed as variadic arguments.
+
+        Parameters:
+            Ts: The types of the variadic arguments.
+
+        Returns:
+            The formated template.
+
+        Variadic arguments are required to implement `Stringable`.
+
+        """
+        var Entries = _FormatCurlyEntry.create_entries(self)
+        var manual_indexing_count = 0
+        var automatic_indexing_count = 0
+        var kwargs_indexing_count = 0
+        for e in Entries:
+            if e[].value.isa[Int]():
+                manual_indexing_count += 1
+            elif e[].value.isa[NoneType]():
+                automatic_indexing_count += 1
+            elif e[].value.isa[String]():
+                kwargs_indexing_count += 1
+
+        if manual_indexing_count and automatic_indexing_count:
+            raise "Cannot both use manual and automatic indexing for *args"
+
+        for e in Entries:
+            if e[].value.isa[String]():
+                raise "Index " + e[].value[String] + " not in kwargs"
+            if manual_indexing_count:
+                if e[].value.isa[Int]() and e[].value[Int] >= len(args):
+                    raise ("Index " + str(e[].value[Int]) + " not in *args")
+            if automatic_indexing_count:
+                if automatic_indexing_count > len(args):
```

Comment:
Since `len` of anything is necessarily non-negative, this check is redundant.
```suggestion
            if automatic_indexing_count > len(args):
```